version_updates.md

```diff
@@ -0,0 +1,120 @@
+# OpenStackVersion
+
+## Overview
+
+This document describes how to use the OpenStackVersion
+CR to apply a minor updates to a deployed OpenStack
+installation (including both Controlplane and Dataplane
+Services)
+
+## Requirements
+
+ * Use of OLM (Operator Lifecycle Manager) is assumed
+ * The CSV (ClusterServiceVersion) for the openstack-operator
+   is used as the primary resource for versioning and the
+   associated ENV variables for OpenStack service containers.
+   These are typically prefixed with RELATED\_IMAGE...
+   The CSV defines the version.
+ * OpenStack operators are always upgraded first via OLM. This
+   can be either via manual or automatic. Upgrading operators
+   will trigger OpenStackVersion to provide access to a new
+   available version.
+ * And ENV called OPENSTACK\_VERSION must be set in the CSV
+   for openstack-operator.  This controls the 'availableVersion'
+   of OpenStackVersion.
+   
+
+## General Update Workflow
+
+OpenStack is installed by following the normal installation procedure. When
+an OpenStackControlplane resource is created in any given namespace an
+OpenStackVersion resource will be immediately created when the OpenStackControlPlane
+is reconciled (1 OpenStackVersion per namespace).
+
+When a new set of Operators is installed via OLM (either manually or automatically)
+the OpenStackVersion resources will reconcile and a new 'AvailableVersion' will
+be set on each CR.
```

Comment:
Container images are never really deleted however if you install a new set of openstack operator's (the entire set) it is very likely that all new image defaults will exist. The old image references will persist on any existing/deployed OpenStackControlplane and will not change or be deleted in anyway until a user kicks off the minor update process. In summary, installing new operators via OLM gives you new image defaults, and the option to perform a minor update as well.
As for manipulations only the image versions are getting patched via OpenStackVersion. Once we make a meaningful CRD change that requires a new version of OpenStackControlplane we'll just have to have conditionals to call the correct reconcile logic across the two of them. Also, I currently have the reconcile code guarded by targetVersion. If a targetVersion has already been reconciled, or is just not set then nothing will happen.

Reply:
> Container images are never really deleted however if you install a new set of openstack operator's (the entire set) it is very likely that all new image defaults will exist. The old image references will persist on any existing/deployed OpenStackControlplane and will not change or be deleted in anyway until a user kicks off the minor update process. In summary, installing new operators via OLM gives you new image defaults, and the option to perform a minor update as well. 

I'm trying to understand how the container images (like nova-api image) is managed in an offline installation. Does the old container images only manually deleted from the cluster after the OpenStackControlPlane is changed to refer only to the new images?

> As for manipulations only the image versions are getting patched via OpenStackVersion. Once we make a meaningful CRD change that requires a new version of OpenStackControlplane we'll just have to have conditionals to call the correct reconcile logic across the two of them. Also, I currently have the reconcile code guarded by targetVersion. If a targetVersion has already been reconciled, or is just not set then nothing will happen.

If I understand your answer correctly then. We allow the human to change OpenStackControlplane after a new set of operators are deployed via OLM but before the image version are changed via OpenStackVersion.  The reconciliation of such human change on OpenStackControlPlane CR will be done by the new operators but still deploying the old container images. I see the following problematic scenario.
1. greenfield deployed with a nova container version X and nova-operator version A
2. a new nova container version Y is built which contains a nova upstream change that require a tweak in the nova configuration generated by the nova-operator
3. a nova-operator version B is built that supports nova container version Y
4. a minor update is started in the deployment. The nova-operator version B is deployed by OLM but the nova container version Y is not used yet as OpenStackVersion.TargetVersion is not bumped to the new AvailableVersion
5. the human do a reconfiguration of OpenStackControlPlane CR, unrelated to the minor update, that results in the deployments of new nova containers.  As the reconciliation of such change already handled by the new nova-operator version B but still using the old nova container image X the nova-operator needs not to do the config tweak as that is only needed by nova container image Y not X. 
Do you also think this is a valid scenario? If so how should nova-operator handle this? Does nova-operator can look at the OpenStackVersion CR to see that the TargetVersion is still not bumped and therefore skip the logic that only required by the new TargetVersion?

Reply:
> I'm trying to understand how the container images (like nova-api image) is managed in an offline installation. Does the old container images only manually deleted from the cluster after the OpenStackControlPlane is changed to refer only to the new images?

In an offline installation container images get mirrored like this: https://docs.openshift.com/container-platform/4.14/installing/disconnected_install/installing-mirroring-installation-images.html
Nothing that is happening here changes that. When you install a new set of operators (each time) you'd need to mirror things (operators, bundles, and all of the associated images RELATED_IMAGES...).

> If I understand your answer correctly then. We allow the human to change OpenStackControlplane after a new set of operators are deployed via OLM but before the image version are changed via OpenStackVersion. The reconciliation of such human change on OpenStackControlPlane CR will be done by the new operators but still deploying the old container images. I see the following problematic scenario.
> 
> 1. greenfield deployed with a nova container version X and nova-operator version A
> 2. a new nova container version Y is built which contains a nova upstream change that require a tweak in the nova configuration generated by the nova-operator
> 3. a nova-operator version B is built that supports nova container version Y
> 4. a minor update is started in the deployment. The nova-operator version B is deployed by OLM but the nova container version Y is not used yet as OpenStackVersion.TargetVersion is not bumped to the new AvailableVersion
> 5. the human do a reconfiguration of OpenStackControlPlane CR, unrelated to the minor update, that results in the deployments of new nova containers.  As the reconciliation of such change already handled by the new nova-operator version B but still using the old nova container image X the nova-operator needs not to do the config tweak as that is only needed by nova container image Y not X.
>    Do you also think this is a valid scenario? If so how should nova-operator handle this? Does nova-operator can look at the OpenStackVersion CR to see that the TargetVersion is still not bumped and therefore skip the logic that only required by the new TargetVersion?

If we need logic to handle complex upgrade scenarios like this I think checking the OpenStackVersion CR's deployed targetVersion and "gating" or adding a conditional around new service operator code could be a valid pattern. Is the example you give here more likely to occur during an upgrade though (instead of a minor update where no new configuration changes exist)?

Reply:
> In an offline installation container images get mirrored like this: https://docs.openshift.com/container-platform/4.14/installing/disconnected_install/installing-mirroring-installation-images.html
> Nothing that is happening here changes that. When you install a new set of operators (each time) you'd need to mirror things (operators, bundles, and all of the associated images RELATED_IMAGES...).

Thanks! I guess similarly the cleanup of the old unused container images done manually.

> If we need logic to handle complex upgrade scenarios like this I think checking the OpenStackVersion CR's deployed targetVersion and "gating" or adding a conditional around new service operator code could be a valid pattern. Is the example you give here more likely to occur during an upgrade though (instead of a minor update where no new configuration changes exist)?

OK. I think we are on the same page here about the OpenStackVersion usage. You are right that we should not do such complex changes in a minor update.